pysonos/discovery.py

```diff
@@ -156,17 +156,30 @@ def create_socket(interface_addr=None):
                 if b"Sonos" not in data:
                     continue
 
-                # pylint: disable=not-callable
-                zone = config.SOCO_CLASS(addr[0])
-                if zone in seen:
+                # Players can be discovered on multiple interfaces and
+                # we do not want to requery the same player once per interface
+                # if we recently polled it in this discovery
+                if addr[0] in seen:
                     continue
 
-                seen.add(zone)
+                # pylint: disable=not-callable
+                soco = config.SOCO_CLASS(addr[0])
+
+                if include_invisible:
+                    zones = soco.all_zones
```

Comment:
> Your experience is not typical. Instability like that is usually indicative of a networking issue, i.e packets being dropped due to firewall rules or NAT rewrites.
> 
> It is stable now but I am not sure that tells us much since you are no longer querying the speaker. So the root cause likely remains.

The affected device is the only one I have on WiFi.  The curious part is that I'm also monitoring it via ICMP and there are no drops.

Reply:
I'll unplug it an see how long it takes to drop

Reply:
It took ~3 minutes to drop out of the Home Assistant UI after unplugging it while running this code. 

The sonos app still sees it and hasn't dropped it yet

Reply:
I'll run the test again to make sure I didn't just get lucky with the timing

Reply:
No network errors when looking http://192.168.106.12:1400/support/review

```
contents of /proc/ath_rincon/status
Debug info for INFRA mode at 17179695

Mode: INFRA (sonosnet)
Operating on channel 2437
Home channel is 2437
RF Chains: RX:3 TX:3
RF Chainmask: RX:0x07 TX:0x07
Max Spatial Streams: RX:3 TX:3
Noise Floor: -110 dBm (chain 0 ctl)
Noise Floor: -110 dBm (chain 1 ctl)
Noise Floor: -109 dBm (chain 2 ctl)
PHY errors since last reading/reset: 72820       
OFDM ANI level: 4
Ch11 Spur Immunity Level: 0
HAL Reset Failures: cnt: 0 last: 0
Region: 1

Node 54:2A:1B:E2:CE:C9 - FROM 41 : TO 46 : STP 03 : MODEL 24.1: KEY 8
  41

Node 54:2A:1B:E2:CF:20 - FROM 36 : TO 39 : STP 03 : MODEL 24.1: KEY 7
  36

Node 54:2A:1B:E3:9F:BB - FROM 41 : TO 44 : STP 03 : MODEL 24.1: KEY 6
  41

Node 54:2A:1B:E3:9F:D6 - FROM 38 : TO 42 : STP 03 : MODEL 24.1: KEY 5
  38
```
```

running /sbin/ifconfigath0      Link encap:Ethernet  HWaddr 34:7E:5C:50:2A:D1             inet6 addr: fe80::367e:5cff:fe50:2ad1%1995797040/64 Scope:Link           UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1           RX packets:0 errors:0 dropped:0 overruns:0 frame:0           TX packets:0 errors:0 dropped:0 overruns:0 carrier:0           collisions:0 txqueuelen:199            RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)           Interrupt:155 Memory:c0e00000-c0e20000
--



```

Reply:
2nd test took ~4 minutes.

Both times the sonos app took much longer for the device to disappear

Reply:
`root@mauinasa:/volume1/docker/89-homeassistant# ping -f 192.168.106.12
PING 192.168.106.12 (192.168.106.12) 56(84) bytes of data.
.^C
--- 192.168.106.12 ping statistics ---
2563 packets transmitted, 2562 received, 0.0390168% packet loss, time 584ms
rtt min/avg/max/mdev = 0.637/0.956/13.784/0.836 ms, pipe 2, ipg/ewma 1.007/0.730 ms`

ping test shows no loss.

Reply:
The support review page shows `Office` as a `Tertiary Node`

Reply:
It looks like the cache expires after 5 seconds which should be fine since the discovery period is 60s.

`        self._zgs_cache = Cache(default_timeout=5)`

```
        zgs = self.zoneGroupTopology.GetZoneGroupState(cache=self._zgs_cache)[
            "ZoneGroupState"
        ]
```

Maybe this changed?

Reply:
So more testing:

The speaker _always_ responds to SSDP requests when its playing.
Once it goes idle for a bit it only responds intermittently.

I wonder if it has some type of sleep mode that causes it to miss the SSDP request?

Reply:
I've replaced the access point and the Ethernet. Still drops after a bit of idle

Reply:
Let's assume it's the device itself since I've ruled everything else out. I ordered a new Sonos one.

Reply:
Is pysonos and the speaker on the same subnet with no proxies, routers or firewalls?

Reply:
Same subnet, same network, no proxies, no firewall.

All the sonos devices are on the same subnet.

Reply:
> The affected device is the only one I have on WiFi.

What does this mean exactly? Sonos should build its SonosNet mesh if you have some wired and some wireless speakers but that is not what your screenshot shows.

Reply:
The wired speakers are powered off currently.

Reply:
I turned on all 8 (6 wired, 2 wireless). Only the One has the missing SSDP responses issue.

Behavior is still the same.

Reply:
I'm going to drop two of the wired ones and see how long they take to go unavailable

Reply:
It took ~9m for the hard wired ones to go unavailable almost exactly the same time they disappeared from the sonos app.

Reply:
Going to drop the last 4 wired and see what happens

Reply:
At this point the one is not responding to ssdp, but it is pinging just fine and a ping flood show no loss.

Reply:
The affected device is hardware version `1.21.1.8-1.1`. Its much older than all the other ones

Reply:
Running software 13.0 build 62186220 (both the move -- which is fine) and the (one -- affected device)

Reply:
Took 19 minutes to go unavailable.. almost exactly the same time as the sonos app.

Reply:
I think that's the extent of the testing I can do.  Will report back when the new hardware arrives

Reply:
When wiring multiple Sonos speakers you need to be careful with its creation of SonosNet and network loops. It also only works with classic STP, not RSTP. Best practices say to disable the wifi on all but one of the wired speakers.